plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/rest/TrinoRestCatalog.java

```diff
@@ -134,22 +136,34 @@ public TrinoRestCatalog(
     @Override
     public boolean namespaceExists(ConnectorSession session, String namespace)
     {
-        return restSessionCatalog.namespaceExists(convert(session), Namespace.of(namespace));
+        return restSessionCatalog.namespaceExists(convert(session), toNamespace(namespace));
     }
 
     @Override
     public List<String> listNamespaces(ConnectorSession session)
     {
-        return restSessionCatalog.listNamespaces(convert(session)).stream()
-                .map(Namespace::toString)
-                .collect(toImmutableList());
+        return collectNamespaces(session, Namespace.empty());
+    }
+
+    private List<String> collectNamespaces(ConnectorSession session, Namespace parentNamespace)
+    {
+        ArrayList<String> namespacesList = new ArrayList<>();
+
+        List<Namespace> childNamespaces = restSessionCatalog.listNamespaces(convert(session), parentNamespace);
+
+        for (Namespace childNamespace : childNamespaces) {
+            namespacesList.add(childNamespace.toString());
+            namespacesList.addAll(collectNamespaces(session, childNamespace));
+        }
+
+        return namespacesList;
     }
 
     @Override
     public void dropNamespace(ConnectorSession session, String namespace)
     {
         try {
-            restSessionCatalog.dropNamespace(convert(session), Namespace.of(namespace));
+            restSessionCatalog.dropNamespace(convert(session), toNamespace(namespace));
```

Comment:
we should list and delete all the child namespaces and then the requested namespace.

Reply:
@mayankvadariya From my perspective, it's unnecessary to handle underlying namespaces automatically. These namespaces may contain other objects that could be deleted without explicit user notification. Therefore, I believe that any management operations, particularly those involving data deletion, should remain solely under the user's control and responsibility.

Reply:
Makes sense, I might have tested cascade and commented here. There is a cascade command `DROP SCHEMA archive CASCADE` https://trino.io/docs/current/sql/drop-schema.html#examples which allows dropping underlying objects. 

```sql
trino:tpch> create schema iceberg."level1";
CREATE SCHEMA
trino:tpch> create schema iceberg."level1.level2";
CREATE SCHEMA
trino:tpch> create schema iceberg."level1.level2.level3";
CREATE SCHEMA
trino:tpch> show schemas like '%level%';
        Schema        
----------------------
 level1               
 level1.level2        
 level1.level2.level3 
(3 rows)

Query 20240924_220453_00029_xqmmv, FINISHED, 3 nodes
Splits: 17 total, 17 done (100.00%)
0.88 [5 rows, 86B] [5 rows/s, 98B/s]

trino:tpch> drop schema iceberg.level1;
Query 20240924_220511_00030_xqmmv failed: Failed to drop namespace: level1

```

Reply:
That reminds me of whether to support recursively creating a ns. 

```sql
trino:tpch> create schema iceberg."level1.level2";
Query 20240925_021526_00028_wkf4w failed: Namespace does not exist: level1

trino:tpch> create schema iceberg.level1;
CREATE SCHEMA
trino:tpch> create schema iceberg."level1.level2";
CREATE SCHEMA
```

Reply:
For example, Spark doesn't create or delete namespaces recursively. Each of these actions results in the following exceptions:
```
Exception in thread "main" org.apache.iceberg.exceptions.NoSuchNamespaceException: Namespace does not exist: aaa
```

Deletion also fails even with CASCADE (and there is no mention in the documentation that CASCADE will delete non-empty namespaces):
```
Exception in thread "main" org.apache.iceberg.exceptions.BadRequestException: Malformed request: Namespace aaa is not empty
```

Personally, I would prefer to leave this behaviour as-is, as it keeps things clear.

Reply:
The current implementation in this PR is wrong. The dropped schema definitely shouldn't appear. 
```sql
CREATE SCHEMA "a";
CREATE SCHEMA "a.b";

CREATE TABLE "a".test_a(a int);
CREATE TABLE "a.b".test_b(a int);

DROP SCHEMA "a" CASCADE; 
SHOW SCHEMAS;

DROP SCHEMA
       Schema
--------------------
 a
 a.b
 information_schema
 tpch
```

Reply:
@ebyhr I disagree with you because:
- Iceberg does not support cascading on the DROP SCHEMA query. The schema must be empty for a successful deletion, as outlined in the [REST Catalog API specification](https://github.com/apache/iceberg/blob/main/open-api/rest-catalog-open-api.yaml).

<img width="622" alt="Screenshot 2024-09-30 at 10 22 36 PM" src="https://github.com/user-attachments/assets/587ece96-ebf1-4e6d-88b9-e5387df58aa1">

- The `TrinoRestCatalog.dropNamespace()` method calls `restSessionCatalog.dropNamespace()` but ignores its return value, which would be **false** in the situation you described.
- The underlying Iceberg catalog's `dropNamespace()` method does not return a list of tables within the namespace, leading to an incorrect `NoSuchNamespaceException`. I'll raise an issue to address this.